### PR TITLE
Support custom encodings on Windows through GNU libiconv

### PR DESCRIPTION
**Pull request by HertzDevil**
This PR adds support for `IO`s with custom encodings on Windows, using https://github.com/pffang/libiconv-for-Windows, which wraps GNU libiconv in an MSVC project.

To make things work more generically across platforms, there is now a new flag `-Dwithout_iconv` which disables both libiconv and the built-in `iconv_*` functions from the C library, plus all the associated tests; any attempt to use custom encodings will raise during runtime, similar to Windows before this PR. This is useful when the target platform has a poor iconv implementation (e.g. Android NDK) and libiconv is not yet available on that platform. The code below is all that is needed to select between libiconv and iconv at the moment:

```crystal
# src/crystal/iconv.cr
{% if flag?(:win32) %}
  require "./lib_iconv"
  private USE_LIBICONV = true
{% else %}
  require "c/iconv"
  private USE_LIBICONV = false
{% end %}
```

Thus Windows is hardcoded to use libiconv, and other platforms will continue to use iconv, but it is easy to change this in the future. If `-Dwithout_iconv` is provided then `src/crystal/iconv.cr` itself would not be required.

The decision to distinguish `LibIconv` from `LibC` is intentional, because they are not strictly API-compatible (e.g. FreeBSD has `__iconv`, GNU libiconv has `iconvctl`).

Related: #5430

### CODE REVIEW

File: src/crystal/iconv.cr [hunk 1]
```diff
@@ -1,9 +1,21 @@
-require "c/iconv"
+{% if flag?(:win32) %}
```
Comment (oprypin) [inline src/crystal/iconv.cr:1]: Wasn't there a flag `:msvc`? If you're saying this is specific to MSVC

Comment (HertzDevil) [reply src/crystal/iconv.cr:1]: Maybe, but to date nobody seems to really attempt building Crystal on Cygwin or MSYS(2), so to my understanding `flag?(:msvc)` is currently reserved for code that actually invokes the MSVC build tools.

**Review by straight-shoota — APPROVED** [review]

File: .github/workflows/win.yml [hunk 2]
```diff
@@ -75,6 +75,19 @@ jobs:
...
+        uses: actions/checkout@v2
+        with:
+          repository: pffang/libiconv-for-Windows
+          ref: 9b7aba8da6e125ef33912fa4412779279f204003 # master @ 2021-08-24
+          path: libiconv
+      - name: Build libiconv
+        if: steps.cache-libs.outputs.cache-hit != 'true'
+        working-directory: ./libiconv
+        run: |
+          [IO.File]::WriteAllLines("$pwd\libiconv.vcxproj", (Get-Content "libiconv.vcxproj").Replace("MultiThreadedDLL", "MultiThreaded"))
```
Comment (oprypin) [inline .github/workflows/win.yml:89]: Can't this be `sed`? We have one instance already below.

```suggestion
          sed -i 's/MultiThreadedDLL/MultiThreaded/' libiconv.vcxproj
```

**Review by oprypin — APPROVED** [review]

File: .github/workflows/win.yml (hunk 2 above)
Comment (HertzDevil) [reply .github/workflows/win.yml]: I always wondered why that worked on Windows CI and it seems `sed.exe` comes from the Git Bash.

File: src/crystal/lib_iconv.cr [hunk 3]
```diff
@@ -0,0 +1,18 @@
+require "c/stddef"
+
+{% if flag?(:without_iconv) %}
+  {% raise "The `without_iconv` flag is preventing you to use the LibIconv module" %}
```
Comment (Sija) [inline src/crystal/lib_iconv.cr:4]:
```suggestion
  {% raise "The `without_iconv` flag is preventing you from using the LibIconv module" %}
```

File: src/crystal/lib_iconv.cr [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+{% end %}
+
+@[Link("iconv")]
```
Comment (yxhuvud) [inline src/crystal/lib_iconv.cr:7]: Hmm, does this mean everyone that wants to use strings on windows need to have the iconv lib installed? GNU libiconv is LGPL, which allows *dynamic* linking, but not static.

http://www.gnu.org/licenses/gpl-faq.html#LGPLStaticVsDynamic

Comment (HertzDevil) [reply src/crystal/lib_iconv.cr:7]: It turns out MPIR has the same issue. Same goes for GMP on non-Windows platforms actually.

Comment (straight-shoota) [reply src/crystal/lib_iconv.cr:7]: It's not "everyone who want's to use string". If you don't need the encoding support, you can pass `-Dwithout_iconv`. But I guess that doesn't help much.

Comment (HertzDevil) [reply src/crystal/lib_iconv.cr:7]: I think it is fine as long as the compiler itself can be compiled without those LGPL components on Windows (which is already the case), but we have to be very clear about all the third-party software used by Crystal and which standard library types link to them, probably directly under the Crystal book's home.

Comment (oprypin) [reply src/crystal/lib_iconv.cr:7]: Everyone that wants to use strings on Windows will have to provide source code.

This is not a new problem or a problem for this pull request to solve.

Actually maybe it's not even a problem at all. Linking against dynamic libraries works just as well (exact same method, even), just that nobody bothered to set it up that way yet, and the CI builds happen to ship static libraries.

**Review by oprypin — APPROVED** [review]

File: src/crystal/lib_iconv.cr (hunk 4 above)
Comment (straight-shoota) [reply src/crystal/lib_iconv.cr:7]: Let's continue the discussion in https://github.com/crystal-lang/crystal/issues/11488

**Comment by oprypin** [comment]
Every invocation of Crystal now gives a warning
```
iconv.lib(iconv.obj) : warning LNK4099: PDB 'libiconvStatic.pdb' was not found with 'iconv.lib(iconv.obj)' or at 'D:\a\crystal\crystal\libiconvStatic.pdb'; linking object as if no debug info
iconv.lib(localcharset.obj) : warning LNK4099: PDB 'libiconvStatic.pdb' was not found with 'iconv.lib(localcharset.obj)' or at 'D:\a\crystal\crystal\libiconvStatic.pdb'; linking object as if no debug info
```
https://github.com/crystal-lang/crystal/runs/4338209730?check_suite_focus=true#step:37:34

Currently that disrupts compilation of Shards
https://github.com/crystal-lang/install-crystal/runs/4338333591?check_suite_focus=true#step:4:29
https://github.com/crystal-lang/shards/blob/v0.16.0/Makefile#L27

**Pull request merged**